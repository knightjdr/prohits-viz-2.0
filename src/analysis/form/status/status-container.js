import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import getTaskStatus from '../../../state/post/task-thunk';
import Status from './status';
import taskSelector from '../../../state/selectors/task-selector';

class StatusContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      status: 'loading',
    };
  }
  componentWillMount = () => {
    const { fetchTaskStatus, taskID } = this.props;
    fetchTaskStatus(taskID);
  }
  componentWillReceiveProps = (nextProps) => {
    this.updateContent(nextProps.tasks, nextProps.taskID, this.props.tasks);
    this.updateTasks(nextProps.tasks.shouldUpdate);
  }
  updateContent = (tasks, taskID, prevTasks) => {
    const currentStatus = tasks.status.find(task => task.id === taskID);
    const prevStatus = prevTasks.status.find(task => task.id === taskID);
    if (
      currentStatus &&
      (
        this.state.status === 'loading' ||
        currentStatus.status !== prevStatus.status
      )
    ) {
      this.setState({
        status: currentStatus.status,
      });
    }
  }
  updateTasks = (shouldUpdate) => {
    if (shouldUpdate) {
      const { fetchTaskStatus, taskID } = this.props;
      fetchTaskStatus(taskID);
    }
  }
  render() {
    return (
      <Status
        closeModal={this.props.closeModal}
        status={this.state.status}
        taskID={this.props.taskID}
        visible={Boolean(this.props.taskID)}
      />
    );
  }
}

StatusContainer.defaultProps = {
  taskID: null,
};

StatusContainer.propTypes = {
  closeModal: PropTypes.func.isRequired,
  fetchTaskStatus: PropTypes.func.isRequired,
  taskID: PropTypes.string,
  tasks: PropTypes.shape({
    didError: PropTypes.bool,
    isUpdating: PropTypes.bool,
    list: PropTypes.arrayOf(
      PropTypes.string,
    ),
    shouldUpdate: PropTypes.bool,
    status: PropTypes.arrayOf(
      PropTypes.shape({}),
    ),
  }).isRequired,
};

/* istanbul ignore next */
const mapStateToProps = state => ({
  tasks: taskSelector(state),
});

/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchTaskStatus: (id) => {
    dispatch(getTaskStatus(id));
  },
});

const ConnectedContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(StatusContainer);

export default ConnectedContainer;
