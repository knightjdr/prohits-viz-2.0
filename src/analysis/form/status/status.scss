@import 'style/theme';

.analysis__task-status-alert {
  margin-bottom: 10px;
}
.analysis__task-status_default {
  margin-right: 5px;
}
.analysis__task-status_green {
  color: $success;
  margin-right: 5px;
}
.analysis__task-status_red {
  color: $warning;
  margin-right: 5px;
}