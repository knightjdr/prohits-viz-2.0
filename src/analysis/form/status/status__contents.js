import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import PropTypes from 'prop-types';
import React, { Fragment } from 'react';
import {
  faCog,
  faCheck,
  faExclamationTriangle,
} from '@fortawesome/pro-solid-svg-icons';

import Link from '../../../components/router-link/router-link';

const statusContent = {
  complete: {
    alert: 'The task is complete.',
    color: 'analysis__task-status_green',
    icon: faCheck,
    link: true,
    message: (
      <p>
        Click on the link below to view the results.
      </p>
    ),
    spin: false,
  },
  error: {
    alert: 'There was an error running the task.',
    color: 'analysis__task-status_red',
    icon: faExclamationTriangle,
    link: true,
    message: (
      <p>
        Click on the link below to view the error log.
      </p>
    ),
    spin: false,
  },
  loading: {
    alert: 'Initializing...',
    color: 'analysis__task-status_default',
    icon: faCog,
    link: null,
    spin: true,
  },
  running: {
    alert: 'The task is running.',
    color: 'analysis__task-status_default',
    icon: faCog,
    link: true,
    message: (
      <Fragment>
        <p>You can track its status from:</p>
        <ol>
          <li>This window</li>
          <li>From the link below</li>
          <li>
            By clicking the &quot;Tasks&quot; link on the navigation menu
            at the top of the page
          </li>
        </ol>
        <p>The task results will be available for 24 hours.</p>
      </Fragment>
    ),
    spin: true,
  },
};

const StatusContent = ({
  status,
  id,
}) => {
  const content = statusContent[status];
  const link = content.link ? `/visualization/${id}` : null;
  return (
    <Fragment>
      <div className="analysis__task-status-alert">
        <FontAwesomeIcon
          className={content.color}
          icon={content.icon}
          spin={content.spin}
        />
        {content.alert}
      </div>
      {content.message}
      {
        link &&
        <Link to={link}>
          {process.env.REACT_APP_HOME_ROOT}{link}
        </Link>
      }
    </Fragment>
  );
};

StatusContent.defaultProps = {
  id: PropTypes.string,
};

StatusContent.propTypes = {
  id: PropTypes.string,
  status: PropTypes.string.isRequired,
};

export default StatusContent;
