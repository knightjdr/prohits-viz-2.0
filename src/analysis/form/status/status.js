import PropTypes from 'prop-types';
import React from 'react';
import { Button, Modal } from 'antd';

import Content from './status__contents';

import './status.css';

const Status = ({
  closeModal,
  status,
  taskID,
  visible,
}) => (
  <Modal
    closable={false}
    footer={[
      <Button key="close" onClick={closeModal} type="primary">Close</Button>,
    ]}
    onCancel={closeModal}
    okText="Close"
    title={`Task ID: ${taskID}`}
    visible={visible}
  >
    <div className="analysis__task-status">
      <Content
        id={taskID}
        status={status}
      />
    </div>
  </Modal>
);

Status.defaultProps = {
  taskID: null,
};

Status.propTypes = {
  closeModal: PropTypes.func.isRequired,
  status: PropTypes.string.isRequired,
  taskID: PropTypes.string,
  visible: PropTypes.bool.isRequired,
};

export default Status;
