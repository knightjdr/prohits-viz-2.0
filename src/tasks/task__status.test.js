import { shallow } from 'enzyme';

import status from './task__status';

describe('Task status', () => {
  describe('complete', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = shallow(
        status('complete'),
      );
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should have correct class', () => {
      expect(wrapper.hasClass('tasks__complete')).toBeTruthy();
    });
  });

  describe('error', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = shallow(
        status('error'),
      );
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should have correct class', () => {
      expect(wrapper.hasClass('tasks__error')).toBeTruthy();
    });
  });

  describe('default status is running', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = shallow(
        status('other'),
      );
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should have correct class', () => {
      expect(wrapper.hasClass('tasks__running')).toBeTruthy();
    });
  });
});
