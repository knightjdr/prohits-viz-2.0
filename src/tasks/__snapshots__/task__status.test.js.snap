// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task status complete should match snapshot 1`] = `
<span
  className="tasks__complete"
>
  <FontAwesomeIcon$1
    border={false}
    className=""
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f00c",
          "M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z",
        ],
        "iconName": "check",
        "prefix": "fas",
      }
    }
    listItem={false}
    mask={null}
    name=""
    pull={null}
    pulse={false}
    rotation={null}
    size={null}
    spin={false}
    symbol={false}
    transform={null}
  />
  complete
</span>
`;

exports[`Task status default status is running should match snapshot 1`] = `
<span
  className="tasks__running"
>
  <FontAwesomeIcon$1
    border={false}
    className=""
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f110",
          "M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z",
        ],
        "iconName": "spinner",
        "prefix": "fas",
      }
    }
    listItem={false}
    mask={null}
    name=""
    pull={null}
    pulse={false}
    rotation={null}
    size={null}
    spin={true}
    symbol={false}
    transform={null}
  />
  running
</span>
`;

exports[`Task status error should match snapshot 1`] = `
<span
  className="tasks__error"
>
  <FontAwesomeIcon$1
    border={false}
    className=""
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          576,
          512,
          Array [],
          "f071",
          "M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z",
        ],
        "iconName": "exclamation-triangle",
        "prefix": "fas",
      }
    }
    listItem={false}
    mask={null}
    name=""
    pull={null}
    pulse={false}
    rotation={null}
    size={null}
    spin={false}
    symbol={false}
    transform={null}
  />
  error
</span>
`;
